Source/Dafny/Verifier/Translator.ExpressionWellformed.cs

```diff
@@ -1139,45 +1139,45 @@ private class WFOptions {
       BoogieStmtListBuilder builder, ExpressionTranslator etran) {
       // If we're inside an "old" expression, then "etran" will know how to translate
       // expressions. However, here, we're also having to translate e.S, which is a
-      // Statement. Since statement translation (in particular, CallStmt's) works
-      // directly on the global variable $Heap, we change its value here.
-      // (Note, this would be a bad translation if the Statement could introduce
-      // control flow that would miss the "$Heap := tmpHeap;" assignment that
-      // resets the value of $Heap. Luckily, StmtExpr does not allow any such
-      // Statement.)
-
-      void addStatements() {
-        TrStmt(stmtExpr.S, builder, locals, etran);
-      }
-
-      void withFreshHeap(string heapVarSuffix, IToken token) {
-        var suffix = CurrentIdGenerator.FreshId(heapVarSuffix);
-        var tmpHeapVar = new Bpl.LocalVariable(token, new Bpl.TypedIdent(token, "Heap$" + suffix, predef.HeapType));
-        locals.Add(tmpHeapVar);
-        var tmpHeap = new Bpl.IdentifierExpr(token, tmpHeapVar);
-        var generalEtran = new ExpressionTranslator(this, predef, token);
-        var theHeap = generalEtran.HeapCastToIdentifierExpr;
-
-        // tmpHeap := $Heap;
-        builder.Add(Bpl.Cmd.SimpleAssign(token, tmpHeap, theHeap));
-        // $Heap := etran.$Heap;
-        builder.Add(Bpl.Cmd.SimpleAssign(token, theHeap, etran.HeapExpr));
-
-        addStatements();
-
-        // $Heap := tmpHeap;
-        builder.Add(Bpl.Cmd.SimpleAssign(token, theHeap, tmpHeap));
-      }
-
+      // Statement. Since statement translation (in particular, translation of CallStmt's)
+      // work directly on the global variable $Heap, we temporarily change its value here.
       if (etran.UsesOldHeap) {
-        withFreshHeap("StmtExpr#", stmtExpr.S.Tok);
+        BuildWithHeapAs(stmtExpr.S.Tok, "StmtExpr#", locals, builder,
+          etran, () => TrStmt(stmtExpr.S, builder, locals, etran));
       } else {
-        addStatements();
+        TrStmt(stmtExpr.S, builder, locals, etran);
       }
 
       CheckWellformedWithResult(stmtExpr.E, options, result, resultType, locals, builder, etran);
     }
 
+    /// <summary>
+    /// Temporarily set the heap to the value etran.HeapExpr and call build().
+    ///
+    /// The Boogie code generated by build() should have just one exit point, namely at the
+    /// end of the code generated. (Any other exit would cause control flow to miss
+    /// BuildWithHeapAs's assignment that restores the value of $Heap.)
+    /// </summary>
+    void BuildWithHeapAs(IToken token, string heapVarSuffix, List<Variable> locals,
+      BoogieStmtListBuilder builder, ExpressionTranslator etran, System.Action build) {
```

Comment:
It's less reusable when passing `ExpressionTranslator` instead of `Bpl.Expr`, but alright ^^

Reply:
Ah, good idea. I had missed that. Since I had to change whitespace formatting elsewhere, I also improved this.